File: lib/models/recommendation.dart
import 'place.dart';
import 'place_image.dart';

class Recommendation {
  const Recommendation({
    required this.id,
    required this.name,
    required this.category,
    required this.latitude,
    required this.longitude,
    required this.reviewCount,
    required this.isPopular,
    required this.isHeritage,
    required this.isLocalSpeciality,
    required this.matchedCategories,
    required this.recommendationScore,
    this.rating,
    this.recommendationReason,
    this.accessConfidence = 'PUBLIC_LIKELY',
    this.isSaved = false,
    this.normalizedCategory = 'other',
    this.image,
  });

  final String id;
  final String name;
  final String category;
  final double latitude;
  final double longitude;
  final double? rating;
  final int reviewCount;
  final bool isPopular;
  final bool isHeritage;
  final bool isLocalSpeciality;
  final List<PlaceCategory> matchedCategories;
  final double recommendationScore;
  final String? recommendationReason;
  final String accessConfidence;
  final bool isSaved;
  final String normalizedCategory;
  final PlaceImageData? image;

  factory Recommendation.fromJson(Map<String, dynamic> json) {
    final rawMatchedCategories =
        json['matched_categories'] as List<dynamic>? ?? const [];
    final imageJson = json['image'];
    return Recommendation(
      id: json['id'] as String,
      name: json['name'] as String,
      category: json['category'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      rating: (json['rating'] as num?)?.toDouble(),
      reviewCount: json['review_count'] as int,
      isPopular: json['is_popular'] as bool,
      isHeritage: json['is_heritage'] as bool,
      isLocalSpeciality: json['is_local_speciality'] as bool,
      matchedCategories: rawMatchedCategories
          .map((value) => PlaceCategory.values.byName(value as String))
          .toList(growable: false),
      recommendationScore: (json['recommendation_score'] as num).toDouble(),
      recommendationReason: json['recommendation_reason'] as String?,
      accessConfidence:
          (json['access_confidence'] as String?) ?? 'PUBLIC_LIKELY',
      isSaved: (json['is_saved'] as bool?) ?? false,
      normalizedCategory:
          json['normalized_category'] as String? ??
          json['category'] as String? ??
          'other',
      image: imageJson is Map
          ? PlaceImageData.fromJson(Map<String, dynamic>.from(imageJson))
          : null,
    );
  }
}

File: backend/app/schemas/recommendation.py
"""Validation schemas for multi-category place recommendations."""

from base64 import urlsafe_b64decode
from binascii import Error as BinasciiError

from enum import Enum
from uuid import UUID

from pydantic import field_validator
from sqlmodel import Field, SQLModel

from app.schemas.place_image import PlaceImageRead


class DiscoveryCategory(str, Enum):
    RELIGIOUS = "religious"
    FOOD = "food"
    TOURISM = "tourism"
    CAFES = "cafes"
    HERITAGE = "heritage"
    MARKETS = "markets"
    NATURE = "nature"


class RecommendationRequest(SQLModel):
    categories: list[DiscoveryCategory] = Field(min_length=1, max_length=7)
    limit: int = Field(default=30, ge=1, le=100)
    trip_id: UUID | None = None
    purposes: list[str] | None = None
    interests: list[str] | None = None
    category_filter: DiscoveryCategory | None = None
    cursor: str | None = Field(default=None, max_length=32)

    @field_validator("categories")
    @classmethod
    def remove_duplicate_categories(
        cls,
        categories: list[DiscoveryCategory],
    ) -> list[DiscoveryCategory]:
        return list(dict.fromkeys(categories))

    @field_validator("cursor")
    @classmethod
    def validate_cursor(cls, cursor: str | None) -> str | None:
        if cursor is None:
            return None
        try:
            decoded = urlsafe_b64decode(cursor.encode("ascii")).decode("ascii")
            if int(decoded) < 0:
                raise ValueError
        except (BinasciiError, UnicodeError, ValueError) as exc:
            raise ValueError("cursor is invalid") from exc
        return cursor

    @property
    def cursor_offset(self) -> int:
        if self.cursor is None:
            return 0
        return int(urlsafe_b64decode(self.cursor.encode("ascii")).decode("ascii"))


class RecommendationRead(SQLModel):
    id: UUID
    name: str
    category: str
    latitude: float
    longitude: float
    rating: float | None
    review_count: int
    is_popular: bool
    is_heritage: bool
    is_local_speciality: bool
    matched_categories: list[DiscoveryCategory]
    recommendation_score: float
    recommendation_reason: str | None = None
    access_confidence: str = "PUBLIC_LIKELY"
    is_saved: bool = False
    normalized_category: str = "other"
    image: PlaceImageRead | None = None
